src/table_properties_collector.rs

```diff
@@ -0,0 +1,90 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crocksdb_ffi::{self, DBEntryType, DBUserCollectedProperties, DBTablePropertiesCollector};
+use libc::{c_void, c_char, c_int, uint8_t, uint64_t, size_t};
+use std::collections::HashMap;
+use std::mem;
+use std::slice;
+
+/// `TablePropertiesCollector` provides the mechanism for users to collect
+/// their own properties that they are interested in. This class is essentially
+/// a collection of callback functions that will be invoked during table
+/// building. It is constructed with TablePropertiesCollectorFactory. The methods
+/// don't need to be thread-safe, as we will create exactly one
+/// TablePropertiesCollector object per table and then call it sequentially
+pub trait TablePropertiesCollector {
+    /// The name of the properties collector.
+    fn name(&self) -> &str;
+
+    /// Will be called when a new key/value pair is inserted into the table.
+    fn add_userkey(&mut self, key: &[u8], value: &[u8], entry_type: DBEntryType);
+
+    /// Will be called when a table has already been built and is ready for
+    /// writing the properties block.
+    fn finish(&mut self) -> HashMap<Vec<u8>, Vec<u8>>;
```

Comment:
Since `DB::get_property_int` already uses `&str` as prop key,
I'd prefer `&str` as key here too.
Not sure if it is suitable here.